### PR TITLE
[Merged by Bors] - feat(RingTheory/Flat/FaithfullyFlat): definition of faithfully flat module

### PR DESCRIPTION
**Pull request by jjaassoonn**
1. $M$ is $R$-faithfully flat iff $M$ is flat and $IM \ne M$ for all maximal ideal $I \le R$
2. prove that this is equivalent to that $M$ is flat and $M \otimes N \ne 0$ whenever $N \ne 0$

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+-/
+@[mk_iff] class FaithfullyFlat : Prop where
+  flat : Module.Flat R M := by infer_instance
+  submodule_ne_top :  ∀ ⦃m : Ideal R⦄ (_ : Ideal.IsMaximal m), m • (⊤ : Submodule R M) ≠ ⊤
+
+namespace FaithfullyFlat
+
+attribute [instance] FaithfullyFlat.flat
+
+instance self (R : Type u) [CommRing R] : FaithfullyFlat R R where
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/Flat/FaithfullyFlat.lean:52]: I think `R` is already in scope.

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+end proper_ideal
+
+section faithful
+
+attribute [-simp] Ideal.Quotient.mk_eq_mk in
+lemma iff_flat_and_rTensor_faithful :
+    FaithfullyFlat R M ↔
+    (Flat R M ∧
+      ∀ (N : Type max u v) [AddCommGroup N] [Module R N],
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/Flat/FaithfullyFlat.lean:74]: Maybe you can use `∀ ⦃N : Type ...'⦄`, but as you prefer.

Does this hold with `N` in an arbitrary universe? (Of course we only want the `→` direction with such an `N`.)

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+namespace Module
+
+variable (R : Type u) (M : Type v) [CommRing R] [AddCommGroup M] [Module R M]
+
+/--
+A module `M` over a commutative ring `R` is *faithfully flat* if it is flat and,
+for all `R`-module homomorphism `f : N → N'` such that `id ⊗ f = 0`, we have `f = 0`.
+-/
+@[mk_iff] class FaithfullyFlat : Prop where
```
Comment (mbkybky) [inline Mathlib/RingTheory/Flat/FaithfullyFlat.lean:44]: I think maybe `FaithfullyFlat` can be defined by extending `Module.Flat`.

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  refine ⟨fun ⟨flat, h⟩ => ⟨flat, fun I hI r => ?_⟩, fun h => ⟨h.1, fun m hm => h.2 _ hm.ne_top⟩⟩
+  obtain ⟨m, hm, le⟩ := I.exists_le_maximal hI
+  exact h hm <| eq_top_iff.2 <| show ⊤ ≤ m • ⊤ from r ▸ Submodule.smul_mono le (by simp [r])
+
+end proper_ideal
+
+section faithful
+
+attribute [-simp] Ideal.Quotient.mk_eq_mk in
+lemma iff_flat_and_rTensor_faithful :
```
Comment (mbkybky) [inline Mathlib/RingTheory/Flat/FaithfullyFlat.lean:71]: I think you can add a theorem that states `FaithfullyFlat R M` implies `∀ (N : Type*) [AddCommGroup N] [Module R N], Nontrivial N → Nontrivial (N ⊗[R] M))`. This maybe be convenient for practice.

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean (hunk 3 above)
Comment (jjaassoonn) [reply Mathlib/RingTheory/Flat/FaithfullyFlat.lean:44]: Yes! good idea

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean (hunk 4 above)
Comment (jjaassoonn) [reply Mathlib/RingTheory/Flat/FaithfullyFlat.lean:106]: I have added instance `Module.FaithfullyFlat.lTensor_nontrivial` and `Module.FaithfullyFlat.rTensor_nontrivial`

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean (hunk 2 above)
Comment (jjaassoonn) [reply Mathlib/RingTheory/Flat/FaithfullyFlat.lean:109]: I have added instance `Module.FaithfullyFlat.lTensor_nontrivial` that works for `N` in arbitrary `universes`

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  haveI H : Subsingleton (M ⧸ m • (⊤ : Submodule R M)) := by
+    rwa [Submodule.subsingleton_quotient_iff_eq_top]
+  rw [← not_nontrivial_iff_subsingleton] at H
+  contradiction
+
+lemma iff_flat_and_rTensor_reflects_triviality :
+    FaithfullyFlat R M ↔
+    (Flat R M ∧
+      ∀ (N : Type max u v) [AddCommGroup N] [Module R N],
+        Subsingleton (N ⊗[R] M) → Subsingleton N) :=
```
Comment (mbkybky) [inline Mathlib/RingTheory/Flat/FaithfullyFlat.lean:131]: Maybe you can also add a theorem that states `FaithfullyFlat R M` implies `∀ (N : Type*) [AddCommGroup N] [Module R N], Subsingleton (N ⊗[R] M) → Subsingleton N`.

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+namespace Module
+
+variable (R : Type u) (M : Type v) [CommRing R] [AddCommGroup M] [Module R M]
+
+/--
+A module `M` over a commutative ring `R` is *faithfully flat* if it is flat and,
+for all `R`-module homomorphism `f : N → N'` such that `id ⊗ f = 0`, we have `f = 0`.
+-/
+@[mk_iff] class FaithfullyFlat extends Module.Flat R M : Prop where
+  submodule_ne_top :  ∀ ⦃m : Ideal R⦄ (_ : Ideal.IsMaximal m), m • (⊤ : Submodule R M) ≠ ⊤
```
Comment (mbkybky) [inline Mathlib/RingTheory/Flat/FaithfullyFlat.lean:45]:
```suggestion
  submodule_ne_top : ∀ ⦃m : Ideal R⦄ (_ : Ideal.IsMaximal m), m • (⊤ : Submodule R M) ≠ ⊤
```

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+namespace FaithfullyFlat
+
+attribute [instance] FaithfullyFlat.toFlat
```
Comment (mbkybky) [inline Mathlib/RingTheory/Flat/FaithfullyFlat.lean:49]:
```suggestion

```
Now this instance will be automatically generated, so this line is unnecessary.

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean (hunk 5 above)
Comment (jjaassoonn) [reply Mathlib/RingTheory/Flat/FaithfullyFlat.lean:148]: done, `Module.FaithfullyFlat.(r)lTensor_reflects_triviality`

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean (hunk 2 above)
Comment (jjaassoonn) [reply Mathlib/RingTheory/Flat/FaithfullyFlat.lean:122]: For the `⦃N : Type ...'⦄`, I think an explicit `N` works better. When dealing with exactness, i.e. `⦃N1 N2 N3 : Type ...'⦄`, I think I will use `⦃⦄`

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    [FaithfullyFlat R M] (N : Type*) [AddCommGroup N] [Module R N]
+    [h : Subsingleton (N ⊗[R] M)] : Subsingleton N := by
+  revert h; change _ → _; contrapose
+  simp only [not_subsingleton_iff_nontrivial]
+  intro h
+  infer_instance
+
+lemma lTensor_reflects_triviality
+    [FaithfullyFlat R M] (N : Type*) [AddCommGroup N] [Module R N]
+    [Subsingleton (M ⊗[R] N)]:
```
Comment (jcommelin) [inline Mathlib/RingTheory/Flat/FaithfullyFlat.lean:108]:
```suggestion
    [Subsingleton (M ⊗[R] N)] :
```

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    Subsingleton N := by
+  haveI : Subsingleton (N ⊗[R] M) := (TensorProduct.comm R N M).toEquiv.injective.subsingleton
+  apply rTensor_reflects_triviality R M
+
+instance lTensor_nontrivial
+    [FaithfullyFlat R M] (N : Type*) [AddCommGroup N] [Module R N] [Nontrivial N] :
+    Nontrivial (M ⊗[R] N) :=
+  TensorProduct.comm R M N |>.toEquiv.nontrivial
```
Comment (jcommelin) [inline Mathlib/RingTheory/Flat/FaithfullyFlat.lean:116]: Can you please move this to be directly after `rTensor_nontrivial`?

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+/--
+A module `M` over a commutative ring `R` is *faithfully flat* if it is flat and,
+for all `R`-module homomorphism `f : N → N'` such that `id ⊗ f = 0`, we have `f = 0`.
+-/
+@[mk_iff] class FaithfullyFlat extends Module.Flat R M : Prop where
+  submodule_ne_top : ∀ ⦃m : Ideal R⦄ (_ : Ideal.IsMaximal m), m • (⊤ : Submodule R M) ≠ ⊤
+
+namespace FaithfullyFlat
+
+
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/Flat/FaithfullyFlat.lean:49]:
```suggestion

```

**Review by riccardobrasca — COMMENTED** [review]
There is still a minor comment from Johan, but LGTM. Thanks!

bors d+

File: Mathlib/RingTheory/Flat/FaithfullyFlat.lean (hunk 9 above)
Comment (jjaassoonn) [reply Mathlib/RingTheory/Flat/FaithfullyFlat.lean]: done

**Comment by riccardobrasca** [comment]
I think this is the one making bors failing. Can you please merge master, fix the error and merge again?

**Comment by jjaassoonn** [comment]
This should work now

**Pull request closed without merging**